cryptomite/raz.py

```diff
@@ -0,0 +1,466 @@
+
+        Parameters
+        ----------
+        input1 : list of bits (0s and 1s)
+            The first input, consisting of n_1 bits.
+        input2 : list of bits (0s and 1s)
+            The second input, consisting of n_2 < n_1/2 bits.
+
+        Returns
+        -------
+        list of bits (0s and 1s)
+            The extractor output bits, of length m.
+        """
+        assert len(input1) >= 2*self.n
+        assert 0 < len(input2) <= self.n
+
+        x1, x2 = input1[0:self.n], input1[self.n:]
+        y = input2 + [0] * (len(input2) - self.n)
+
+        pad_amount = self.pad_amount
+        x1 = x1 + [0]*pad_amount
+        x2 = x2 + [0]*pad_amount
+        y = y + [0]*pad_amount
+
+        cur_delta = self.gf_mul(y, x1)
+        product = self.__gf_add_one(cur_delta)  # GF 1 times lambda_0
+        cur_delta = self.gf_mul(cur_delta, cur_delta)  # delta^2
+
+        # First loop is starting from product = product * lambda_1,
+        # i.e. (delta^(2^1) + 1)), followed by cur_delta = delta^4
+        # up to product * (lambda_{logp-1}, i.e. delta^{2^(logp-1))
+        # and cur_delta = delta^(2^logp)
+        # Note: we could unwrap the final multiplication at self.logp-1
+        # to save the unnecessary calculation of delta^(2^logp),
+        # but it happens in parallel anyway and makes the code harder to read.
+        for _i in range(1, self.logp):
+            # product times lambda_1 ... lambda_{l-1}
+            (product, cur_delta) = self.ntt.raz_iteration(product, cur_delta,
+                                                          self.n, self.s)
+        conv_output = self.gf_mul(product, x2)
+        return cast(BitsT, [conv_output[i] for i in range(self.n)])
+
+# ------- UTILITY FUNCTIONS -------
+
+
+def log2_error_raz(n_1: int,
+                   k_1: float,
+                   k_2: float,
+                   m: int,
+                   l: int,
+                   p: int):
+    """
+    Compute an upper bound on the logarithm base 2
+    of the error for the efficient weak version of
+    Raz's extractor presented in [Fore2025]_.
+
+    Parameters
+    ----------
+    n_1 : int
+        The length of the first input (in bits).
+    k_1 : float
+        The min-entropy of the first input.
+    k_2 : float
+        The min-entropy of the second input.
+    m : int
+        The length of the extractor output (in bits).
+    l : int
+        The logarithm base 2 of p'.
+        **Note:** the efficient construction requires p'
+        to be a power of 2, i.e. l is an integer.
+    p : int
+        A free parameter that is an even integer that
+        satisfies p <= 2^l / m.
+
+    Returns
+    -------
+    float :
+        An upper bound on the logarithm base 2 of the
+        extractor error.
+    """
+    log_gamma_bound = (n_1 - k_1)/p + max((l - n_1/2 + 1)/p,
+                                          np.log2(p) - k_2/2) + 1
+    return log_gamma_bound + m/2
+
+
+def opt_error_raz(n_1: int,
```

Comment:
This function could maybe do with a few comments or new-lines just to break it up and make it a bit more readable.

Reply:
This has now been updated by Cam